pivotal-tracker/pivotal_integration.php

```diff
@@ -0,0 +1,131 @@
+    }
+}
+
+/**
+ * Do git operations to find all commits between the specified commit hashes,
+ * and return an associative array containing all applicable commits that
+ * contain references to Pivotal Tracker issues.
+ */
+function _get_commits($current_commithash, $last_commithash, $env) {
+    $commits = array(
+        // Raw output of git log since the last processed
+        'history_raw' => NULL,
+        // Formatted array of commits being sent to pivotal
+        'history' => array(),
+        // An array keyed by pivotal ticket id, each holding an
+        // array of commit ids.
+        'pivotal' => array()
+    );
+    //builds command
+    $cmd = 'git log'; // add -p to include diff
+    if (!$last_commithash) {
+        $cmd .= ' -n 1';
+    }
+    else {
+        $cmd .= ' ' . $last_commithash . '...' . $current_commithash;
+    }
+    $commits['history_raw'] = shell_exec($cmd);
+    // Parse raw history into an array of commits
+    $history = preg_split('/^commit /m', $commits['history_raw'], -1, PREG_SPLIT_NO_EMPTY);
+    foreach ($history as $str) {
+        $commit = array(
+            'full' => 'Commit: ' . $str
+        );
+        // Only interested in the lines before the diff now
+        $lines = explode("\n", $str);
+        $commit['id'] = $lines[0];
+        $commit['message'] = trim(implode("\n", array_slice($lines, 4)));
+        $commit['formatted'] = $commit['message'] . " " . $lines[1] . ' - ' . $lines[2];
+        // Look for matches in commit based on Pivotal Tracker issue ID format
+        // Expected pattern: "[#12345678] commit message" or "[fixed|completed|finished #12345678 message]"
+        preg_match('/\[[^#]*#\K\d{1,16}/', $commit['message'], $matches);
+        if (count($matches) > 0) {
+            // Build the $commits['pivotal'] array so there is
+            // only 1 item per ticket id
+            foreach ($matches as $ticket_id) {
+                if (!isset($commits['pivotal'][$ticket_id])) {
+                    $commits['pivotal'][$ticket_id] = array();
+                }
+                // ... and only 1 item per commit id
+                $commits['pivotal'][$ticket_id][$commit['id']] = $commit['id'];
+            }
+            // Add the commit to the history array since there was a match.
+            $commits['history'][$commit['id']] = $commit;
+        }
+    }
+    return $commits;
+}
+
+/**
+ * Send commits to Pivotal
+ */
+function send_commit($secrets, $commit) {
+    print("\n==== Posting to Pivotal Tracker ====\n");
+
+    // Sends commit to pivotal.
+    $ch = 'curl -X POST -H "X-TrackerToken: '. $secrets['tracker_token'] . '" -H "Content-Type: application/json" -d \'{"source_commit":{"commit_id":"' . substr($commit['id'], 0, 10) . '","message":"' . $commit['formatted'] . '"}\' "https://www.pivotaltracker.com/services/v5/source_commits?fields=%3Adefault%2Ccomments"';
+
+    $result = shell_exec($ch);
+    print("RESULT: $result");
```

Comment:
I presume the result does not divulge the tracker_token, or anything else sensitive?